esp8266: Download Espressif RTOS SDK as a new RIOT PKG

iosabi opened the pull request:
### Contribution description

RIOT-OS uses part of Espressif ESP8266 RTOS SDK to build support for
this CPU. The SDK includes some vendor-provided closed source
pre-compiled libraries that we need to modify to adapt to RIOT-OS
usage. This library modifications was done once and uploaded to a fork
of the vendor repository and was provided as an environment variable.

This patch changes two things:

1. It installs the SDK as a RIOT PKG from the new pkg/esp8266_sdk
directory instead of requiring the user to download it separately.

2. It performs the library modifications (symbol renames) on the pkg
Makefile removing the need to use a fork with the modifications applied
and simplifying the SDK update and future modifications.

This change sets the SDK package version (git SHA) to the same one that
our fork was using as a parent in the vendor repository, meaning that
the output libraries are exactly the same as before. 

### Testing procedure
Tested with
```
ESP8266_RTOS_SDK_DIR=/dev/null USEMODULE=esp_log_startup make -C tests/shell BOARD=esp8266-esp-12x flash
```

and verified that the program works. The boot message now includes:
```
ESP8266-RTOS-SDK Version v3.1-51-g913a06a9
```
confirming the SDK version used.

`/dev/null` in the test is just to make sure that no evaluation of
`ESP8266_RTOS_SDK_DIR` in make is affected by the environment variable
value which would be set to the SDK for people who followed the set up
instructions before this change.

Tested the checkout size:
```bash
$ du -hs build/pkg/esp8266_sdk/
124M	build/pkg/esp8266_sdk/
```

### Issues/PRs references

This is relate to issue #16402 and would simplify updating the SDK to fix it.

aabadie commented on pkg/esp8266_sdk/Makefile:
```diff
@@ -0,0 +1,40 @@
+PKG_LICENSE=Apache-2.0
+
+include $(RIOTBASE)/pkg/pkg.mk
+
+ifneq ($(CPU),esp8266)
+  $(error This package can only be used with the ESP8266 CPU)
+endif
```
Instead of this, you could add a Makefile.dep file to the package with the following content:

```mk
# This package can only be used with the ESP8266 CPU
FEATURES_REQUIRED += arch_esp8266
```

iosabi replied:
Done.

gschorcht commented on pkg/esp8266_sdk/Makefile:
```diff
@@ -0,0 +1,40 @@
+
+# This directory is shared across all apps since there's no need to replicate
+# these libraries on each app.
+ESP_SDK_BUILD_DIR = $(PKG_SOURCE_DIR)/build-libs
```
I'm wondering whether it would make sense to keep this variable overridable.
```suggestion
ESP_SDK_BUILD_DIR ?= $(PKG_SOURCE_DIR)/build-libs
```
This might be helpful during development work when upgrading the RIOT port to a newer SDK version.

iosabi replied:
This being a pkg Makefile means that you can't override it with a `ESP_SDK_BUILD_DIR = /some/path/` in your app Makefile unless that's also `export`ed to the submake. However, if this is for development you can call `make ESP_SDK_BUILD_DIR=/some/path` (with the assignment as a make parameter) and that would override this anyway everywhere.

I renamed this to `ESP8266_SDK_BUILD_DIR` and made it a bit more public adding it to the Makefile.include file with a `?=` there. So:

 *  calling `make ESP8266_SDK_BUILD_DIR=$(pwd)/mytestdir` builds the libraries into `mytestdir` instead of the default `build/pkg/esp8266_sdk/build-libs` and also uses that directory as the source.
 * If instead you call `ESP8266_SDK_BUILD_DIR=$(pwd)/mytestdir make` (with the value in ENV instead of a make arg) this will be ignored by the submake because of the `=` but it won't be ignored by the top level make which uses `?=` from the `Makefile.include`.
 
Does this fit your use case? I'm assuming you want to play with symbol renaming in new libs manually and feed those to the link stage of the app. I would personally do that by adding entries to the `symbol_renames.txt` file but that works too.

gschorcht commented on pkg/esp8266_sdk/Makefile:
```diff
@@ -0,0 +1,40 @@
+	$(Q)echo "#define IDF_VER \"$(shell git -C $(PKG_SOURCE_DIR) describe --tags)\"" \
+	  > $@
+
+$(ESP_SDK_BUILD_DIR)/lib%.a: $(PKG_SOURCE_DIR)/components/esp8266/lib/lib%.a | $(ESP_SDK_BUILD_DIR)
```
:+1: nice solution

gschorcht reviewed (commented):
In a second step, the SDK should then also be removed from `riotdocker` as soon as this PR is part of the official release. However, the question then is how do we deal with older RIOT releases. As far as I know, we still don`t have a riotdocker versioning that is compatible with RIOT versioning.

gschorcht commented:
> Tested with
> 
> ```
> ESP8266_RTOS_SDK_DIR=/dev/null USEMODULE=esp_log_startup make -C tests/shell BOARD=esp8266-esp-12x flash
> ```

Shouldn't it work without `ESP8266_RTOS_SDK_DIR=/dev/null` as well? Without this definition, the compilation fails:
```
[INFO] patch esp8266_sdk
"make" -C ..../pkg/esp8266_sdk
"make" -C ..../boards/esp8266-esp-12x
"make" -C ..../boards/common/esp8266
"make" -C ..../core
"make" -C ..../cpu/esp8266
cc1: error: /components/esp8266/include/internal: No such file or directory [-Werror=missing-include-dirs]
```
It seems that `ESP8266_RTOS_SDK_DIR` isn't set when entering `cpu/esp8266` with a fresh SDK package. Once, the SDK package is compiled with this definition, compilation also works without the definition.

iosabi commented:
> Shouldn't it work without `ESP8266_RTOS_SDK_DIR=/dev/null` as well? Without this definition, the compilation fails:

That's an interesting issue. I'll dig deeper into what's up with this. The /dev/null was to make sure that it didn't use the ENV in any way (I deleted my /opt/esp/ESP8266_RTOS_SDK directory). I think I'll to export that var.

iosabi commented:
> That's an interesting issue. I'll dig deeper into what's up with this. The /dev/null was to make sure that it didn't use the ENV in any way (I deleted my /opt/esp/ESP8266_RTOS_SDK directory). I think I'll to export that var.

Indeed, exporting the variable did the trick. I added a comment next to it.

I uploaded a fixup change with all these fixes. Please take a look.

fjmolinas commented:
> In a second step, the SDK should then also be removed from `riotdocker` as soon as this PR is part of the official release. However, the question then is how do we deal with older RIOT releases. As far as I know, we still don`t have a riotdocker versioning that is compatible with RIOT versioning.

As soon as the release is done, we will push https://github.com/RIOT-OS/riotdocker/pull/136 to get versioning in.

MrKevinWeiss commented:
Looks like we have versioning now.

fjmolinas commented:
CI seems happy, do you have more comments @gschorcht? Is there some other tests you would like ran?

iosabi commented:
@fjmolinas Let me know if/when you want me to rebase, squash and test again.
